### PR TITLE
compose: Add types to `createHigherOrderComponent`, `ifCondition` and `pure`

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds types to `createHigherOrderComponent` and `ifCondition`. I converted both to TypeScript as it simplified the typing process greatly for these highly abstract functions.

Part of #18838

## How has this been tested?
Type checks pass. No runtime changes.

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by gziolo** [comment]
> I converted both to TypeScript as it simplified the typing process greatly for these highly abstract functions.

Did we confirm that Gutenberg Mobile works with TS files?

**Comment by sarayourfriend** [comment]
What do you mean @gziolo? It looks like the mobile React native e2e tests worked fine, do we need confirmation beyond that this is safe?

**Comment by gziolo** [comment]
There is also https://github.com/wordpress-mobile/gutenberg-mobile that consumes the source code of WordPress packages. We need to ensure it understands TypeScript. As far as I can tell it doesn't use the same Babel config that Gutenberg uses:
https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/babel.config.js
https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/babel.config.js

Well, we can merge this PR, but as an additional step, we will need to help the mobile team to enable TS support in https://github.com/wordpress-mobile/gutenberg-mobile if necessary.

**Comment by sarayourfriend** [comment]
As long as `gutenberg-mobile` consumes the npm packages they should be okay, right? It's just regular JS at that point. It doesn't look like they're doing anything special with the `@wordpress` scope 🤔

**Comment by gziolo** [comment]
> As long as `gutenberg-mobile` consumes the npm packages they should be okay, right? It's just regular JS at that point. It doesn't look like they're doing anything special with the `@wordpress` scope 🤔

They don't do the regular npm consuming 😅 Every production package declares its entry point for React Native, e.g.:
https://github.com/WordPress/gutenberg/blob/619b2fe3f8fab42e5b0ce8d3ab4f9a9b0e64ab1a/packages/block-editor/package.json#L24

They use sources so we need to ensure that their setup is able to work with TypeScript files. It shouldn't be that much work as I can read at:
https://reactnative.dev/docs/typescript

**Comment by sarayourfriend** [comment]
> Every production package declares its entry point for React Native

🤦‍♀️ of course, I completely forgot about this.

Obviously I'm fully on board with helping them get TypeScript support anyway. What's the best way to do this, let it break and fix it later? I'm not sure how to tell how/what way it will break otherwise, but that doesn't feel good to me.

**Comment by gziolo** [comment]
I asked the mobile team in their channel on WordPress Slack to help here (link requires registration at https://make.wordpress.org/chat):
https://wordpress.slack.com/archives/C02RQC4LY/p1618842203015300

**Comment by gziolo** [comment]
This is what I got from @oguzkocer:

> IRC metro’s default React Native Babel config has TypeScript plugin integrated and enabled for quite some time. I think it should “just work”

We can land this PR as soon as CI turns green.

**Comment by oguzkocer** [comment]
> This is what I got from @oguzkocer:

   > IRC metro’s default React Native Babel config has TypeScript plugin integrated and enabled for quite some time. I think it should “just work”

@gziolo Did you maybe mean to ping someone else? I can't think of a time I dealt with ^, but maybe my memory is failing me.

**Comment by gziolo** [comment]
> @gziolo Did you maybe mean to ping someone else? I can't think of a time I dealt with ^, but maybe my memory is failing me.

Ooops, yes. I meant @ceyhun 🙈

**Comment by ceyhun** [comment]
It would be nice to maybe have a companion Gutenberg Mobile PR to run some extra CI checks.

**Comment by sarayourfriend** [comment]
> It would be nice to maybe have a companion Gutenberg Mobile PR to run the some extra CI checks.

Thanks @ceyhun, what would be the best way to do this?

**Comment by gziolo** [comment]
You can open PR similar to what I had in the past: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3249. You only need to point the Gutenberg git submodule to the hash from this branch.

**Comment by dcalhoun** [comment]
@sarayourfriend @gziolo I actually just finished opening said PR in wordpress-mobile/gutenberg-mobile#3403. We'll see what the results are. 😄

**Comment by sarayourfriend** [comment]
Looks like the GB mobile checks are all green 😁

**Comment by gziolo** [comment]
@dcalhoun, thank you for help 💯

File: packages/compose/src/utils/create-higher-order-component/index.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * @return Component class with generated display name assigned.
+ */
+function createHigherOrderComponent<
+	TObviatedProps extends object,
+	TProps extends TObviatedProps
+>(
+	mapComponentToEnhancedComponent: (
+		OriginalComponent: ComponentType< TProps >
+	) => ComponentType< Subtract< TProps, TObviatedProps > >,
```
Comment (jsnajdr) [inline packages/compose/src/utils/create-higher-order-component/index.ts]: Fun fact: The type of `mapComponentToEnhancedComponent` is `HigherOrderComponent<TObviatedProps>`.

File: packages/compose/src/utils/create-higher-order-component/index.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * Given a function mapping a component to an enhanced component and modifier
+ * name, returns the enhanced component augmented with a generated displayName.
+ *
+ * @param mapComponentToEnhancedComponent Function mapping component to enhanced component.
+ * @param modifierName                    Seed name from which to generated display name.
+ *
+ * @return Component class with generated display name assigned.
+ */
+function createHigherOrderComponent<
+	TObviatedProps extends object,
```
Comment (jsnajdr) [inline packages/compose/src/utils/create-higher-order-component/index.ts]: Could this be called `TRemovedProps`? I had to look up the word "obviated" 🙂

Comment (sarayourfriend) [reply packages/compose/src/utils/create-higher-order-component/index.ts]: Yup! Removed is way simpler, I just didn't think of it before 😅

File: packages/compose/src/utils/create-higher-order-component/index.ts (hunk 1 above)
Comment (sarayourfriend) [reply packages/compose/src/utils/create-higher-order-component/index.ts]: While that's true I don't think we can use the type here becuase we can't pass `TProps` down, as the `HOC` type only accepts it on the call signature and we need to grab it from above. But it is a fun fact 😁

File: packages/compose/src/utils/create-higher-order-component/index.ts (hunk 2 above)
Comment (jsnajdr) [reply packages/compose/src/utils/create-higher-order-component/index.ts]: One more idea: making the naming closer to the programmer's thinking, these props are `TInjectedProps`. The HOC injects them into the inner component. The fact that they are removed from the outer component's type is secondary.

File: packages/compose/src/utils/create-higher-order-component/index.ts (hunk 1 above)
Comment (jsnajdr) [reply packages/compose/src/utils/create-higher-order-component/index.ts]: OK, I hoped for a little bit more than a fun fact but yes, now I see that `HigherOrderComponent` doesn't specify all the generic types we need 🙂

**Comment by jsnajdr** [comment]
The `createHigherOrderComponent` is in fact much more generic that whan the `<TProps, TRemovedProps>` specify. The mapping function can transform any React components to any other, where the props don't need to be related at all:
```ts
type HigherOrderComponent< TInnerProps, TOuterProps > = (
  Inner: ComponentType< TInnerProps >
) => ComponentType< TOuterProps >;

function createHigherOrderComponent< TInnerProps, TOuterProps >(
  mapComponent: HigherOrderComponent< TInnerProps, TOuterProps >,
  modifierName: string
): HigherOrderComponent< TInnerProps, TOuterProps > {
  return ( Inner ) => {
    const Outer = mapComponent( Inner );
    const displayName = Inner.displayName || Inner.name || 'Component';
    Outer.displayName = `${ modifierName }(${ displayName })`;
    return Outer;
  };
}
```
What do you think? Isn't this a better type?

I can then define, e.g., the `withI18n` HOC like this:
```tsx
const withI18n = createHigherOrderComponent( function < TProps >(
  Inner: ComponentType< TProps & I18nContextProps >
) {
  return ( props: TProps ) => {
    const i18nProps = useI18n();
    return <Inner { ...props } { ...i18nProps } />;
  };
},
'WithI18n' );
```
The type annotations are quite minimal, I don't need to use any `as` anywhere, and types are correctly inferred.

In an example component that's wrapped with `withI18n`, TypeScript is able to infer that the outer component has only the `name` prop (from `HelloProps`) and not the i18n ones:
Screenshot 2021-04-26 at 14 09 44

I was unable to make `with18n` work with the original `createHigherOrderComponent< TProps, TRemovedProps >` version.

**Comment by jsnajdr** [comment]
There are some obstacles that prevent the new types to be used by `@wordpress/compose` importers:
- missing `types` field in `package.json`
- missing record in top-level `tsconfig.json`
- the package-level `index.js` file is still just JavaScript, not TypeScript, so there is no `index.d.ts` generated in `build-types` and `import type { ... } from '@wordpress/compose'` can't be resolved.

**Comment by sarayourfriend** [comment]
> There are some obstacles that prevent the new types to be used by @wordpress/compose importers:

Indeed, this PR only partially types `compose` so it won't include any of the things you mentioned in your comment until the package is fully typed.

**Comment by sarayourfriend** [comment]
> What do you think? Isn't this a better type?

Yes! It's much simpler, I seem to have over complicated things 🙂

**Comment by sarayourfriend** [comment]
So actually there is a problem with the types you suggested. How would you fully type annotate the `ifCondition` function @jsnajdr? I'm not sure what to say the return type of `ifCondition` is without introducing further generics into `ifCondition` that don't seem to make sense.

**Comment by sirreal** [comment]
> > There are some obstacles that prevent the new types to be used by @wordpress/compose importers:
> 
> Indeed, this PR only partially types `compose` so it won't include any of the things you mentioned in your comment until the package is fully typed.

@sarayourfriend  This point is still important so that the partial typechecking is run as expected:

> missing record in top-level tsconfig.json

**Comment by jsnajdr** [comment]
@sarayourfriend I was able to type `ifCondition` like this:
```tsx
const ifCondition = < TProps, >(
  predicate: ( props: TProps ) => boolean
): HigherOrderComponent< TProps, TProps > =>
  createHigherOrderComponent(
    ( WrappedComponent ) => ( props ) => {
      if ( ! predicate( props ) ) {
        return null;
      }

      return <WrappedComponent { ...props } />;
    },
    'ifCondition'
  );
```

And types for `pure` can be:
```tsx
const pure = createHigherOrderComponent(
  < TProps, >( Wrapped: ComponentType< TProps > ) => {
    return class extends Component< TProps > {
      shouldComponentUpdate( nextProps: TProps ) {
        return ! isShallowEqual( nextProps, this.props );
      }

      render() {
        return <Wrapped { ...this.props } />;
      }
    };
  },
  'pure'
);
```

There are some differences between these two. The return type of `ifCondition( predicate )` can be named -- it's `HigherOrderComponent< TProps, TProps >` -- because there is the `TProps` type parameter. And the `createHigherOrderComponent` call doesn't need any type annotations at all, because everything is inferred.

But for `pure`, I don't know how to name its type:
```tsx
const pure: WhatTheType = createHigherOrderComponent( ... )
```
What should I write in place of `WhatTheType` to make it work? Because the right-hand side of the expression has type annotations, the type is correctly inferred:

Screenshot 2021-04-27 at 14 01 55

But can it be written in terms of `HigherOrderComponent<TProps, TProps>`? I don't know.

**Comment by sarayourfriend** [comment]
> But for pure, I don't know how to name its type:

This is a major problem for type extraction. We need to be able to either explicitly annotate the return type in TS or as a JSDoc `@type` tag 😞

**Comment by jsnajdr** [comment]
With the `HigherOrderComponent< TRemovedProps, Props >` version, I was unable to make types for a `withI18n`-like component work. Here's my doodle:
```tsx
type I18nProps = {
  __: ( s: string ) => string;
};

const i18nProps: I18nProps = {
  __: ( s ) => s
};

const mapI18n = < TProps, >( Inner: ComponentType< TProps & I18nProps > ) => (
  props: TProps
) => <Inner { ...props } { ...i18nProps } />;

const withI18n = createHigherOrderComponent( mapI18n, 'WithI18n' );
```
Here the return type of `createHigherOrderComponent` is not correctly inferred: it's `HigherOrderComponent< object >` instead of `HigherOrderComponent< I18nProps >`.

And I can't pass the `I18nProps` directly as in `createHigherOrderComponent< I18nProps >`. The function has two type parameters, where one would get fixed to `I18nProps` and the other one, `TProps`, would remain generic. I don't know how to cope with that.

**Comment by jsnajdr** [comment]
> This is a major problem for type extraction. We need to be able to either explicitly annotate the return type in TS

Well the type can be written as
```tsx
<TProps>(Inner: ComponentType<TProps>) => ComponentType<TProps>
```
But can it be written in terms of the `HigherOrderComponent` shorthand?

**Comment by sarayourfriend** [comment]
I think it's fine to write it explicitly as the inferred type shows. Alternatively we could add aliases for these types:

```ts
export type HigherOrderComponent< TInnerProps, TOuterProps > = (
	Inner: ComponentType< TInnerProps >
) => ComponentType< TOuterProps >;

export type SimpleHigherOrderComponent = < TProps >(
	Inner: ComponentType< TProps >
) => ComponentType< TProps >;

export type PropRemovingHigherOrderComponent< TRemovedProps > = < TProps >(
	Inner: ComponentType< TProps & TRemovedProps >
) => ComponentType< TProps >;
```

As always *waves hands* names are hard.

**Comment by sarayourfriend** [comment]
I've tried moving forward with this PR but @jsnajdr I think I've run into another snag. Please not the comment at the end of this playground link:

https://www.typescriptlang.org/play?#code/PQKhFgCgAIWhRAHgFwKYCcB2BDANtAE1QAdVMjMBjAS1QGcpZgpqBbYge3WWgCVVslHgDN0HVtADk6AUMkBuFuy48A3tErZWqXAGFsdVABpoAV2Kl0AMWro6PAL7RR4qbg4EDACwVRgwaHpcakxkAFoCajpsACNcVDDMVBQw4KToTA4wmXt0aiFUAjC2Tm4GSBKVaGQAT1JodV1xTiTQgBU61GgnFwlpWWRfSD8wRjgACWoAcy8MaC4idA1mjlbkOg1sTE3TQ2hiMWIN5A5oGK6OGIA3amw0AgA6aCsuQMQtYnjobGhxgHldNUvHcxtAQgArVBCDbUACMAA5tgcOEdoAB3ai4fCXG53LrIWYZVCFZyvAldOGI-aHY6nc77AyGAjVU7k5bKNYPMaggCSwm+vwBhA49AyHB4ONuaCBXSSJOEry2NWgdFINGE+WpKOOs22aK6RGE2FMuB4J2gAANVA4LeivPkvKD6IZQrcscq6KYYsh0IJ1tAAAzOMQSNlcaYhPBa1Ecfls4iMkkhdktMjIJ4AQW2yQ+8VBsaBUTBOq6FuIphktv+gLR9soXmFosyPHrWym+MJGYACjyVRWjZQLnHCZQVmszqZkKC6CVcMrWB5qBrRWyQmgsHg6FyYCBmJBkqUzZ1ftNZug-uhFk0OWmADzQNo8zBJdBdmkmNp-ScYN-a6AAPmgABeaAAAooAASCfF8AC5oGvVN2k6e9H2fH8aQAqAAEpgMAhDVjTDpSBQr911-VF-0UYZ90QQ9qmPABlWdUEmGYMAvK8xzTYDoBQ8iNn-cDICgtD0Dg-C1iI1A+Iw-9sNw+CuKQ4iH34gCqKgA8qlqepyP4BcbkwKZWLPDiMAku8H30jgrkKNTkjQcgNkuSEhAAnjb0gtp+KgQTIOgjBxKU5ApJkv8ADIrNQAy7Nk+SgLw4LQtU2SNMgUAIB3aAAHFqFs7YfmEUwqGQahVmgVhsAsEIpgFUcb1CFlvmzTBgSoEl6sQngtmZBdImXdBQRwbQTBkZAK0wEtAlarZB2ZTqCMa40pm0UISQxAkBXbF88WZSJVVwbAagAOS0VBt1gUEAAEE19CRKuICz2g4eAZvaggnp4KxiqEMrtge4gapTRazVOMg2rm4HOWu26tAqxcBtO7RoBR1G0fR9GGOJZlhq6Xo7QdJrtowXbCCiT4joyM6Lsy2BoCusaJsUhqW0OugNg2htid9e4yYOynce+dnpiSR4xj3KAipKv6NBkPETPYy9zOClCAtfd8H1I9C-z84SHs+uCFfPJX0E+1XRP4j8tfVnWjEgvql1odAkdQODchq7DDdPRXOJZ82X0tzXvxtiiGkgxmsDA6A1egHCErD4SINHSaeGtnj9eC0Do9E2OqIgpPVnsPmKZOs6eLVh59pLl3oAAH1r7OXweQX66kT6hnz63K-Jw7S+RkCLQAEnUcxLBsOxkCE-Ok7OvQDFQLOHcRsusMgiCcIcUDh+L3ua4cLCLTziCI+2a284cKiHE02iqkNY1TVlgQ0CNszTeCtLk6L3ZUCfexZp-5kIFQIcDgkqOOgFYRpRGLTOAGYTxsSWK-ZmXUzBMiavSZENwiACmKtQAAjqYCkKd-7RwACJnBqKCBa0Mdx7k-jwTmv9kD-x5AQOCeloo2Rqi-E2ZsE4QRCH-dqrC3Y+iBg3TAphWDnHQJfdyIFKBy2ft7Y2vsupTzCkcEwut84AHVfQWEKAbMEyhuCgX6H6SQWEHifVCmvCC3kMKRXUII5hwi2EqjEUZOuGQpEyO6GvOSwlwH8OPqgcakcs7IiOHBRxf4QmqHsfQsExD3E8W-kwlhBAs76KqqQD6wVc72JPrxXJhiCkswaNAB4NTokbCcK4rJQEXGpMHKw7o0BgCUTXhfSCDg7bCUkIw1pADJDYQ-oXHggIgHRjoHBdQCoQGeLyEZeQKShFtI8e7bxEi-FzCcCEjRkQrj-nUHUh4jS0kOFvMAY5QSsITJTvBaZ6JqAEkye4rOgIHmdICM2fIXQvAcDRM8zxmJ8DAimpcwcZDozzEwNufw0dJASEZFIoGBIixFjZAuIu9UFzbG-psQwTw+TQBqBwUwKpPhvJlOstxMKeTkLqfMScKTzTklsPMNE2wPmbJ8ulAIOkug9UIFCQ6MhoAcIMtwlRr8zZ8oAfxQCtKMRYnRFwAA1iYGIbLewEFWJIM09pMCaqBHcYsdLcVTkFeyAlqCuiaD2AqJYmKNjVjtBgLo+oNUmmZOCXYPADxQh4LS80M54ihDnNAGQMU6UstArqGFtZlS9jRAYVBQMGJeh9H6M4qAXWoGsVAIAA

**Comment by jsnajdr** [comment]
@sarayourfriend The trouble is this definition of `withInstanceId`:
```tsx
type InstanceIdProps = { instanceId: string | number };

const withInstanceId: PropRemovingHOC< InstanceIdProps > = createHigherOrderComponent(
  < TProps >( Wrapped: ComponentType< TProps & InstanceIdProps > ) => {
    return ( props: TProps ) => {
      const instanceId = useInstanceId( Wrapped );
      return <Wrapped { ...props } instanceId={ instanceId } />;
    };
  }
);
```

Now when calling `withInstanceId( C )` where `C` has props type `{ foo, instanceId }`, then `TProps` is inferred to be `{ foo, instanceId }` and the props of `Wrapped` are inferred to be `{ foo, instanceId } & { instanceId }`. This satisfies all the type constraints and the type of the returned component includes `{ instanceId }`.

We hoped that `TProps` would be inferred as `{ foo }` and type of `Wrapped` (which `C` needs to be assignable to) would be inferred as `{ foo, instanceId }`. That would also satisfy the type constraints, while being more minimal. But TypeScript made a different choice 🤷 

I was able to fix it with this implementation:
```tsx
const withInstanceId = createHigherOrderComponent(
  < TProps extends InstanceIdProps >( Wrapped: ComponentType< TProps > ) => {
    return ( props: Omit< TProps, keyof InstanceIdProps > ) => {
      const instanceId = useInstanceId( WrappedComponent );
      return <Wrapped { ...props } instanceId={ instanceId } />;
    };
  }
);
```
It explicitly omits `InstanceIdProps` from the result type, instead of adding them to the input type.

React Redux `connect` types do a very similar `Omit` on the result type:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-redux/index.d.ts#L128

They do a lot of additional magic with the `DistributiveOmit`, `Matching` and `Shared` helpers that I assume solve some edge cases, but the basic idea is the same.

**Comment by sarayourfriend** [comment]
Sweet. In that case can we move forward with this PR in its current state @jsnajdr?

File: packages/compose/src/utils/create-higher-order-component/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * of this is the `pure` HOC which does not change the API surface of the component but
+ * simply modifies the internals.
+ */
+export type HigherOrderComponent< TInnerProps, TOuterProps > = (
+	Inner: ComponentType< TInnerProps >
+) => ComponentType< TOuterProps >;
+
+export type SimpleHigherOrderComponent = < TProps >(
+	Inner: ComponentType< TProps >
+) => ComponentType< TProps >;
```
Comment (jsnajdr) [inline packages/compose/src/utils/create-higher-order-component/index.ts:23]: I wish this type could have a better name (what exactly is simple about it?) or if we could avoid it.

If `HigherOrderComponent` defaulted its second argument:
```ts
HigherOrderComponent< TInnerProps, TOuterProps = TInnerProps >
```
then we could write just
```ts
HigherOrderComponent< TInnerProps >
```
to declare a "simple" HOC.

**Review by jsnajdr — APPROVED** [review]
Let's merge this and continue iterating :+1:

**Comment by jsnajdr** [comment]
> In that case can we move forward with this PR in its current state

Let's do it :rocket:

File: packages/compose/src/utils/create-higher-order-component/index.ts (hunk 3 above)
Comment (sarayourfriend) [reply packages/compose/src/utils/create-higher-order-component/index.ts:23]: I'll do this in the next PR. I agree I hate the names I chose!

Comment (jsnajdr) [reply packages/compose/src/utils/create-higher-order-component/index.ts:23]: OK and thanks for working on this. React HOCs are a really complex TypeScript puzzle. The types for React Redux `connect` are almost 700 lines of hardcore TypeScript. I wish we used other patterns (render props, hooks) where types are more straighforward.

**Pull request merged**